app/src/main/java/com/example/drawingactivity/DialogUtils.kt

```diff
@@ -137,6 +136,183 @@ object DialogUtils {
+            submitArray[1] = (satText.text.toString().toFloat() / 100)
+            submitArray[2] = (valText.text.toString().toFloat() / 100)
+            onSubmit.invoke(submitArray, hexText.text.toString())
+            sheet.dismiss()
+        }
+        val cancelButton = view.findViewById<Button>(R.id.nopeButton)
+        cancelButton.setOnClickListener {
+            onSubmit.invoke(null, null)
+            sheet.dismiss()
+        }
+        sheet.setOnCancelListener {
+            onSubmit.invoke(null, null)
+            sheet.dismiss()
+        }
+        /**
+         * Function that updates all the views in the gradient.
+         * Takes in a color state data class to update views.
+         */
+        fun forceViewUpdates(colorState: ColorUtils.ColorState) {
+            val displaySat = (colorState.saturation * 100).toInt()
+            val displayVal = (colorState.value * 100).toInt()
+            hexText.forceText(colorState.hexCode)
+            hueText.forceText(colorState.hue.toString())
+            satText.forceText(displaySat.toString())
+            valText.forceText(displayVal.toString())
+            redText.forceText(colorState.red.toString())
+            greenText.forceText(colorState.green.toString())
+            blueText.forceText(colorState.blue.toString())
+            huePicker.setHue(colorState.hue)
+            gradientPicker.setHue(colorState.hue)
+            gradientPicker.setSaturationAndValue(colorState.saturation, colorState.value)
+            topColorSquare.setBackgroundColor(parseColor(colorState.hexCode))
+        }
+
+        /**
+         * Clears focus on whatever view the user has currently selected.
+         * Currently only called when the user interacts with the Gradient or Hue Slider.
+         */
+        fun clearFocus() {
+            hexText.clearFocus()
+            hueText.clearFocus()
+            satText.clearFocus()
+            valText.clearFocus()
+            redText.clearFocus()
+            greenText.clearFocus()
+            blueText.clearFocus()
+        }
+        val colorStates = ColorUtils.calculateColorWithHSV(previousColor)
+        forceViewUpdates(colorStates)
+        bottomColorSquare.setBackgroundColor(parseColor(colorStates.hexCode))
+        huePicker?.onHueChanged = {
+            clearFocus()
+            val hsv = FloatArray(3)
+            hsv[0] = it.toFloat()
+            hsv[1] = (satText.text.toString().toFloat() / 100)
+            hsv[2] = (valText.text.toString().toFloat() / 100)
+            val calculatedColorState = ColorUtils.calculateColorWithHSV(hsv)
+            gradientPicker.setHue(it)
+            forceViewUpdates(calculatedColorState)
+        }
+        gradientPicker?.onSatValChanged = { saturation, value ->
+            clearFocus()
+            val hsv = FloatArray(3)
+            hsv[0] = (hueText.text.toString().toFloat())
+            hsv[1] = saturation
+            hsv[2] = 1f - value
```

Comment:
why not have this be done when its passing in the value? If all other locations give the actual value this becomes a bug no? Because then those areas passing the "correct" value are getting a wrong output and the one spot passing a correct value is getting the right output. Though maybe this is like a semantics thing?

Reply:
Good catch, I don't validate text inputs in this case and it causes a crash. For most of the other views that address it this way, there is logic to catch if a user inputs something that isn't an int and just ignores the input until they remove it. But if you set it in one view then use another it crashes. Will require some rework for input validation.